### PR TITLE
TAP-603: Improve Mutex usage to fix the concurrency issue

### PR DESCRIPTION
**Pull request by sula-sa**
Currently if multiple threads enter `getCredentials()` at the same time when the token is expired, all of them will fire refresh logic. Because each of them get their own copy of `Tokens` object (using `getLatestTokens()` method) and then they all pass the same outdated value into the if statements that are below it. To fix this we need to put mutex a bit higher in the logic and put `getCredentials()` inside the mutex block, this way we ensure that whoever enter the block will get the most recent state of our `Tokens`

This change passes the test created by @stoyicker in https://github.com/tidal-music/tidal-sdk-android/pull/25, however, we decided to not proceed with synchronized logic but to improve mutex usage

### CODE REVIEW

**Review by stoyicker — DISMISSED** [review]
lgtm but if possible let's wait until Michael looks at it before merging please

**Comment by sula-sa** [comment]
Created a test that calls `getCredentials()` at the same time with 100 threads. With our previous implementation, the test fails and makes as many refresh calls as there are threads. With the correct mutex position, there is only one refresh call

**Review by stoyicker — APPROVED** [review]

**Review by stoyicker — DISMISSED** [review]

File: auth/src/test/kotlin/com/tidal/sdk/auth/TokenRepositoryTest.kt
Comment (michpohl) [inline auth/src/test/kotlin/com/tidal/sdk/auth/TokenRepositoryTest.kt:761]: There is another test that this one most likely replaces: 
```
`A call to getCredentials should not trigger a backend call if one is already in progress`
```
That test should now be removed I guess.

**Review by michpohl — CHANGES_REQUESTED** [review]

**Comment by michpohl** [comment]
Thanks for taking care of this @stoyicker  and @sula-sa!

File: auth/src/test/kotlin/com/tidal/sdk/auth/TokenRepositoryTest.kt
Comment (sula-sa) [reply auth/src/test/kotlin/com/tidal/sdk/auth/TokenRepositoryTest.kt:720]: Updated

**Comment by sula-sa** [comment]
> Thanks for taking care of this @stoyicker and @sula-sa!

Huge thanks to @stoyicker , that's mainly him, not me :)

**Comment by stoyicker** [comment]
> > Thanks for taking care of this @stoyicker and @sula-sa!
> 
> Huge thanks to @stoyicker , that's mainly him, not me :)

Me who did absolutely nothing: sure! 😎

**Comment by michpohl** [comment]
once this in, we should issue a new release. Can you do that, @sula-sa ?

**Comment by sula-sa** [comment]
> once this is in, we should issue a new release, and use it in the client. Can you do that, @sula-sa ?

Sure, i'll take care of it

**Comment by sula-sa** [comment]
just need your approval @michpohl :)

**Review by michpohl — APPROVED** [review]
🚀

**Pull request merged**
